### PR TITLE
Select between GRCh38 and HG19

### PR DESCRIPTION
**Pull request by jpdna**
Allows selection between `GRCh38` and `hg19` using the `--reference-genome` parameter
as discussed in #132

This code has only been tested for parameter parsing, it has not yet been run on a cluster, let me know if you want me to test it further before this PR is accepted

### CODE REVIEW

**Comment by fnothaft** [comment]
LGTM!

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 1]
```diff
@@ -119,8 +118,13 @@ def launch_pipeline(params):
                                '-o', 'UserKnownHostsFile=/dev/null',
                                '-o', 'StrictHostKeyChecking=no',
                                'screen', '-dmS', params.cluster_name])
-        # Run command on screen session
-        
+        # Run command on screen session  
+
+        if params.reference_genome == 'GRCh38':
+            from toil_scripts.adam_uberscript.input_files import GRCh38_inputs as inputs
+        if params.reference_genome == 'hg19':
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: elif

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 2]
```diff
@@ -119,8 +118,13 @@ def launch_pipeline(params):
...
+            from toil_scripts.adam_uberscript.input_files import hg19_inputs as inputs
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: add else branch raising exception?

File: src/toil_scripts/adam_uberscript/input_files.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
 
-inputs = {    "ref": "https://s3-us-west-2.amazonaws.com/cgl-pipeline-inputs/variant_grch38_reordered/GRCh38_full_analysis_set_plus_decoy_hla.reordered.fa",
+GRCh38_inputs = {    "ref": "https://s3-us-west-2.amazonaws.com/cgl-pipeline-inputs/variant_grch38_reordered/GRCh38_full_analysis_set_plus_decoy_hla.reordered.fa",
               "amb": "https://s3-us-west-2.amazonaws.com/cgl-pipeline-inputs/variant_grch38_reordered/GRCh38_full_analysis_set_plus_decoy_hla.reordered.fa.amb",
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/input_files.py]: indentation seems off

File: src/toil_scripts/adam_uberscript/input_files.py [hunk 4]
```diff
@@ -12,3 +12,19 @@
             "dbsnp": "https://s3-us-west-2.amazonaws.com/cgl-pipeline-inputs/variant_grch38_reordered/ALL_20141222.dbSNP142_human_GRCh38.snps.vcf",
              "omni": "https://s3-us-west-2.amazonaws.com/cgl-pipeline-inputs/variant_grch38_reordered/1000G_omni2.5.grch38.reordered.vcf",
            "hapmap": "https://s3-us-west-2.amazonaws.com/cgl-pipeline-inputs/variant_grch38_reordered/hapmap_3.3.grch38.reordered.vcf"}
+
+hg19_inputs = {    "ref": "https://s3-us-west-2.amazonaws.com/cgl-pipeline-inputs/variant_hg19_reordered/hg19.fa",
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/input_files.py]: ditto

File: src/toil_scripts/adam_uberscript/adam_uberscript.py (hunk 2 above)
Comment (fnothaft) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: This gets validated during argument parsing. See https://github.com/BD2KGenomics/toil-scripts/pull/133/files#diff-7060afbea50ff85e89fe57836bc27314R451.

Comment (jpdna) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: > This gets validated during argument parsing. See https://github.com/BD2KGenomics/toil-scripts/pull/133/files#diff-7060afbea50ff85e89fe57836bc27314R451.
@hannes-ucsc - let me know if you still advise changes here.

Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: I would still use elif and have an else: branch that asserts False. That's just basic defensive programming. Of course, *right now* Frank is right, but *over time* his assertion may break.

Comment (fnothaft) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: My strong preference is to validate arguments as early as is possible; my understanding is that by setting the `choices` param when argparsing, we push the validation to when the arguments are parsed and cannot get invalid inputs here. I'd prefer to document this decision in the argparse section of the code, and to not revalidate here.

Also, this is personal preference, but I prefer to `assert` if an internal invariant fails, and to raise an exception (e.g., `IllegalArgumentException` in Java, `ValueError` in Python) if a user provides bad input. I find `assert False` to be a smell; if you're `assert False`ing, you should almost always `raise Exception("informative error message")` instead. In my world, if an `assert` fires, that indicates that there is a bug in the program that should be fixed, while raising an exception to the user tells them that they need to fix their inputs.

Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: Yes, validate early and raise RuntimeException to alert user but **also** assert here. If someone changes the validation code and forgets to change this code here, they will have introduced a bug which would be caught by the assert. This is especially important considering Python's lack of type-safe enumerations. And -1 on informational message, the stack trace will be informative enough.

Comment (fnothaft) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: > Yes, validate early and raise RuntimeException to alert user but also assert here. If someone changes the validation code and forgets to change this code here, they will have introduced a bug which would be caught by the assert.

Just wanted to point out that the flip side of this is that if someone adds a new reference build, then they need to modify validation code in two places. My preference is to validate the inputs in a single place, and to validate them as early as is possible.

> And -1 on informational message, the stack trace will be informative enough.

Are you actually _-1_ on this? It doesn't make sense to me to require the end user to parse out a stack trace when you can raise an informative error message instead.

Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: After talking to @fnothaft, this is the compromise: add `elif` and `else` with `assert False, "unknown reference genome %s" % params.reference_genome`.

**Comment by hannes-ucsc** [comment]
I want clean the überscript up a bit so I will go ahead and rebase and merge this. Otherwise, I'd be causing even more conflicts.

@jpdna I use SourceTree and XCode's FileMerge for conflict resolution. PyCHarm also has an excellent GUI for merge conflict resolution.

**Comment by jpdna** [comment]
I just anyhow  pushed the re-based version to my repo, and so will update
the pull request anyhow, so you can start from wherever you prefer Hannes -
thanks!

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 5]
```diff
@@ -126,29 +125,32 @@ def launch_pipeline(params):
                                '-o', 'StrictHostKeyChecking=no',
                                'screen', '-dmS', params.cluster_name])
 
-        # do we have a defined master ip?
-        masterIP_arg = ""
-        if params.master_ip:
-            masterIP_arg = "--master_ip %s" % params.master_ip
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: Did you intent to remove this?

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 6]
```diff
@@ -126,29 +125,32 @@ def launch_pipeline(params):
...
+                           "--uuid_manifest ~/manifest " + \
+                           "--ref {ref} " + \
+                           "--amb {amb} " + \
+                           "--ann {ann} " + \
+                           "--bwt {bwt} " + \
+                           "--pac {pac} " + \
+                           "--sa {sa} " + \
+                           "--fai {fai} "
+        if 'alt' in inputs:
+          pipeline_command +=   pipeline_command += "--alt {alt} "
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: we use four spaces for indent

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 7]
```diff
@@ -126,29 +125,32 @@ def launch_pipeline(params):
...
-                            "--alt {alt} " +
-                            "--use_bwakit " +
+        # Run command on screen session  
+
+        if params.reference_genome == 'GRCh38':
+            from toil_scripts.adam_uberscript.input_files import GRCh38_inputs as inputs
+        elif params.reference_genome == 'hg19':
+            from toil_scripts.adam_uberscript.input_files import hg19_inputs as inputs
+        else:
+            raise Exception("Invalid reference genome, reference should have been  validated as a choice during argument parsing.  Should not have reached this point, arg parsing code likely out of date.")
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: This is way too verbose and doesn't contain the value that could not be matched.

**Comment by hannes-ucsc** [comment]
Thanks, I'll take it from here.

File: src/toil_scripts/adam_uberscript/adam_uberscript.py (hunk 6 above)
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: And it's got and extra `pipeline_command += ` in there

Comment (jpdna) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: Indeed, I'll check more carefully, thanks Hannes

**Comment by hannes-ucsc** [comment]
Superceded by #159

**Pull request closed without merging**